### PR TITLE
Delete patient tests

### PR DESCRIPTION
**Pull request by sacuribraimo**
- Added actionsComponent for back, transfer, view, edit and delete action buttons

- Updated to include transfer patient between modules test

- Refactored the delete patient tests considering that a patient should be deleted only from the registration dashboard page.

### CODE REVIEW

File: test/e2eTests/pages/windows/deletePatientModal.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    this.I.say(`${LOG_TAG} Verifying that the save button is disabled`);
+    this.I.seeElement(locate('button').withAttr({ disabled: 'disabled' }).withText(this.translate('SAVE')));
+  }
+
+  verifyDeletionImpossibility() {
+    this.I.say(`${LOG_TAG} Verifying that it is not possible to delete the patient`);
+    const alertElement = '.alert-info';
+    this.I.waitForElement(alertElement, 10);
+    this.I.see(this.translate('COMMON_CANT_DELETE_PATIENT_WITH_HISTORY'), alertElement);
+    this.I.seeElement(locate('button').withAttr({ name: 'isDead', disabled: 'disabled' }));
```
Comment (drryanjames) [inline test/e2eTests/pages/windows/deletePatientModal.js]: See above comment

File: test/e2eTests/pages/windows/deletePatientModal.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  }
+
+  fillDeleteReason(reason) {
+    this.I.say(`${LOG_TAG} filling in the reason for deleting the patient`);
+    this.I.fillField(this.fields.deleteReason, reason);
+  }
+
+  verifySaveButtonIsDisabled() {
+    this.I.say(`${LOG_TAG} Verifying that the save button is disabled`);
+    this.I.seeElement(locate('button').withAttr({ disabled: 'disabled' }).withText(this.translate('SAVE')));
```
[the remark this answers is not in the document]
Comment (drryanjames) [reply test/e2eTests/pages/windows/deletePatientModal.js]: with variable names that clearly state what's happening.

File: test/e2eTests/pages/windows/modal.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+//
+class Modal {
+  constructor(options) {
+    // Initializing the buttons array
+    options.buttons = options.buttons || [];
+    this.options = options;
+
+    this.modalDialog = '.modal-dialog';
+
+    // if no buttons are provided add the cancel and save buttons
```
Comment (drryanjames) [inline test/e2eTests/pages/windows/modal.js]: Try using defaults instead. Check these out:

https://stackoverflow.com/questions/31342290/es6-classes-default-value

https://gist.github.com/ericelliott/f3c2a53a1d4100539f71

File: test/e2eTests/pages/windows/deletePatientModal.js [hunk 4]
```diff
@@ -0,0 +1,74 @@
+const Modal = require('./modal');
+
+const LOG_TAG = '[DeletePatientModal]';
+
+class DeletePatientModal extends Modal {
```
Comment (drryanjames) [inline test/e2eTests/pages/windows/deletePatientModal.js]: This feels more like a component. Models are part of the page just like other components, so not making this a component feels odd. It removes it from the existing page-component architecture, which might make explaining and maintaining the architecture more difficult. I'm about to write documentation and this feels a bit out of place."

What do you think?

File: test/e2eTests/pages/windows/deletePatientModal.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (sacuribraimo) [reply test/e2eTests/pages/windows/deletePatientModal.js]: This is part of the CodeceptJs locator. I'm not sure if it is possible to split it and call .withAttr() for example on an element.

File: test/e2eTests/pages/windows/deletePatientModal.js (hunk 4 above)
Comment (sacuribraimo) [reply test/e2eTests/pages/windows/deletePatientModal.js]: I started with that approach but felt the modal wasn't part of a page but rather was an independent object...some sort of "page" that is part of a main page. Weird! Let's see how angular defines it.

File: test/e2eTests/pages/windows/modal.js (hunk 3 above)
Comment (sacuribraimo) [reply test/e2eTests/pages/windows/modal.js]: Will do! Thanks!

**Comment by sacuribraimo** [comment]
@drryanjames I'm going to make some changes to these tests. It turns out, the delete patient functionality should be from the registration module only.

**Comment by sacuribraimo** [comment]
@drryanjames This is ready for review.

File: test/e2eTests/pages/components/actionsComponent.js [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    this._clickButton('edit', this.actionButtons.editPatient);
+    const editPatientPage = require('./../editPatientPage');
+    editPatientPage._init();
+    editPatientPage.isLoaded();
+    return editPatientPage;
+  }
+
+  clickDeletePatientButton() {
+    this._clickButton('delete', this.actionButtons.deletePatient);
+    // const deletePatientModal = require('./../windows/deletePatientModal.js');
```
Comment (drryanjames) [inline test/e2eTests/pages/components/actionsComponent.js]: Delete these comments if they're not needed.

File: test/e2eTests/pages/components/deletePatientModalComponent.js [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    this.I.selectOption(this.fields.deathReason, deathReason);
+    this.I.click(this.fields.deathDate);
+    this.I.fillField(this.fields.deathDate, deathDate);
+    this.I.click('Done');
+    this._clickSaveButton();
+  }
+
+  verifyRestrictionToDeletePatient() {
+    const alertElement = '.alert-info';
+    const disabledCheckbox = locate('button').withAttr({ name: 'isDead', disabled: 'disabled' });
```
Comment (drryanjames) [inline test/e2eTests/pages/components/deletePatientModalComponent.js]: :)

File: test/e2eTests/pages/components/deletePatientModalComponent.js [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    this.I.say(`${LOG_TAG} Verifying patient death details`);
+    const patientHeader = '.patientdeceased';
+    this.I.waitForElement(patientHeader, 10);
+    this.I.see(`${this.translate('PATIENT_IS_DEAD')} ${this.translate('COMMON_AT')}: ${deathDetails.date} ${this.translate('PATIENT_INFO_DEATH_REASON')}: ${deathDetails.reason}`);
+
+    this.I.say(`${LOG_TAG} Verifying the edit and delete buttons are disabled`);
+    const disabledEditButton = locate('button').withAttr({ id: 'edit_patient', disabled: 'disabled' });
+    const disabledDeleteButton = locate('button').withAttr({ id: 'delete_patient', disabled: 'disabled' });
+    this.I.waitForElement('#delete_patient', 10);
+    // this.I.seeElement(disabledEditButton); // There is a bug on the system... this should be disabled
```
Comment (drryanjames) [inline test/e2eTests/pages/components/deletePatientModalComponent.js]: Is there a bug in GitHub? Should you put the bug number in a comment here?

File: test/e2eTests/pages/components/deletePatientModalComponent.js [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+    this.I.seeElement(disabledDeleteButton);
+  }
+
+  _verifySaveButtonIsDisabled() {
+    const disabledSaveButton = locate('button').withAttr({ disabled: 'disabled' }).withText(this.translate('SAVE'));
```
Comment (drryanjames) [inline test/e2eTests/pages/components/deletePatientModalComponent.js]: :)

File: test/e2eTests/pages/registrationDashboardPage.js [hunk 9]
```diff
@@ -10,7 +10,7 @@ class RegistrationDashboardPage extends Page {
         element: '[ui-sref="dashboard.program"]',
         urlPart: '/registration/#/dashboard',
       },
-      components: ['checkIn', 'tabs'],
+      components: ['checkIn', 'tabs', 'actions', 'deletePatientModal'],
```
Comment (drryanjames) [inline test/e2eTests/pages/registrationDashboardPage.js]: I like deletePatientModal as a component. Feels right.

File: test/e2eTests/tests/dashboard_test.js [hunk 10]
```diff
@@ -1,12 +1,54 @@
...
+  I.say(`${LOG_TAG} Search for patient's identifier`);
+  registrationPage.searchForPatientByIdAndSelect(Data.patients.patient2);
+
+  I.say(`${LOG_TAG} Make sure the registration dashboard page is loaded`);
+  RegistrationDashboardPage.isLoaded();
+
+  RegistrationDashboardPage.clickTransferPatientButton();
+
+  const socialDashboardPage = RegistrationDashboardPage.transferToSocialModule();
+  socialDashboardPage.clickTransferPatientButton();
```
Comment (drryanjames) [inline test/e2eTests/tests/dashboard_test.js]: Would be nice to have logs here.

File: test/e2eTests/tests/deletePatient_test.js [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+const LOG_TAG = '[DeletePatientTests]';
+
+let Patient = null;
+
+Before(async (I, Apis, Data) => {
+  I.say(`${LOG_TAG} Creating patient to be deleted`);
+  Patient = await Apis.patient.create(Data.patients.patient2);
+});
+
+// TODO: Centralize this to be reused
```
Comment (drryanjames) [inline test/e2eTests/tests/deletePatient_test.js]: Might as well do this in these changes, right? Maybe we should have a utils class. I'm thinking /tests/e2eTests/utils.js

export default class Utils {
    static cleanUpPatientVisit(patient) { ... }
}

Or something like that

Comment (drryanjames) [reply test/e2eTests/tests/deletePatient_test.js]: And the other functions, of course

File: test/e2eTests/tests/dashboard_test.js (hunk 10 above)
Comment (sacuribraimo) [reply test/e2eTests/tests/dashboard_test.js]: The methods log some messages like "Clicking on the transfer button" and "Transferring the patient to social module"

File: test/e2eTests/tests/deletePatient_test.js (hunk 11 above)
Comment (sacuribraimo) [reply test/e2eTests/tests/deletePatient_test.js]: I'm reviewing the patient registration next and I'll include the utils class.

**Pull request merged**